### PR TITLE
HHH-17634 Merging a new entity having a @GeneratedValue id should not set the generated id to the original entity

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-17634

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java [hunk 1]
```diff
@@ -319,17 +382,23 @@ private void saveTransientEntity(
...
 		final Object entity = event.getEntity();
 		final EventSource source = event.getSession();
 		final EntityPersister persister = source.getEntityPersister( event.getEntityName(), entity );
 		final String entityName = persister.getEntityName();
-
-		Object id = getDetachedEntityId( event, entity, persister );
+		final Object clonedIdentifier;
+		if ( copiedId == null ) {
+			copiedId = persister.getIdentifier( entity, source );
+			clonedIdentifier = persister.getIdentifierType().deepCopy( copiedId, source.getFactory() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java:395]:
```suggestion
		if ( originalId == null ) {
			originalId = persister.getIdentifier( entity, source );
		}
		final Object clonedIdentifier;
		if ( copiedId == null ) {
			clonedIdentifier = persister.getIdentifierType().deepCopy( originalId, source.getFactory() );
```

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java]: done :)

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java [hunk 2]
```diff
@@ -319,17 +382,25 @@ private void saveTransientEntity(
...
 		final EntityPersister persister = source.getEntityPersister( event.getEntityName(), entity );
 		final String entityName = persister.getEntityName();
-
-		Object id = getDetachedEntityId( event, entity, persister );
+		if ( originalId == null ) {
+			originalId = persister.getIdentifier( entity, source );
+		}
+		final Object clonedIdentifier;
+		if ( copiedId == null ) {
+			clonedIdentifier = persister.getIdentifierType().deepCopy( copiedId, source.getFactory() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java:397]: Looks like tests are not covering this case yet because you try to copy `copiedId` which is always null here.

```suggestion
			clonedIdentifier = persister.getIdentifierType().deepCopy( originalId, source.getFactory() );
```

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java [hunk 3]
```diff
@@ -177,30 +219,53 @@ private void merge(MergeEvent event, MergeContext copiedAlready, Object entity)
...
+			EventSource session,
+			MergeContext mergeContext) {
+		if ( originalId != null && persister.getIdentifierType().isComponentType() ) {
+			/*
+			this is needed in case of composite id containing an association with a generated identifier, in such a case
+			generating the EntityKey will cause a NPE when trying to get the hashcode of the null id
+			 */
+			return copyCompositeTypeId( originalId, (CompositeType) persister.getIdentifierType(), session, mergeContext );
+		}
+		return persister.getIdentifierType().deepCopy( originalId, session.getFactory() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java:235]:
```suggestion
		return null;
```

File: hibernate-core/src/main/java/org/hibernate/event/spi/MergeContext.java [hunk 4]
```diff
@@ -75,6 +75,8 @@ public class MergeContext implements Map<Object,Object> {
 	private final EventSource session;
 	private final EntityCopyObserver entityCopyObserver;
 
+	private final Map<Object,Object> copiedCompositeId = new IdentityHashMap<>(10);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/event/spi/MergeContext.java:78]: Please revert the changes here in this file.

**Review by beikov — APPROVED** [review]
Other than the now unnecessary changes in `MergeContext`, this looks good.
